pocs/linux/kernelctf/CVE-2023-6817_lts_cos/docs/exploit.md

```diff
@@ -0,0 +1,238 @@
+    new_setelem_with_chain(socket, table, tmp_set, NULL, 0, key, 0x40, key_end, 0x40, target_chain);
+    struct nlmsghdr **msg_list = malloc(sizeof(struct nlmsghdr *)*2);
+    msg_list[0] = del_setelem_msg(table, tmp_set, key, 0x40, key_end, 0x40);
+    msg_list[1] = del_set_msg(table, tmp_set);
+    send_msg_list(socket, msg_list, 2);
+}
+```
+`primitive_0` can implement `target_object->use--` and `primitive_1` can implement `target_chain->use--`.
+
+### Leak info
+
+This part of the exploit is the basically same as CVE-2023-4569. If you want to understand more details, read this [article](https://github.com/google/security-research/tree/master/pocs/linux/kernelctf/CVE-2023-4569_lts/docs/exploit.md).
+
+I leak some useful infomation by the following steps.
+
+- 1. Create many objects first.(`object A`,`object B`,...)
+- 2. Create 0xa4 set elements wihch use one of the objects created in step 1. We assume we use `object F`. Create another set element `element X` using `object F`(This set element will be used in step 10). After step2, we will set `object F->use = 0xa5`
+- 3. Call `primitive_0` 0xa5 times, finally making `object F->use = 0`
+- 4. Delete `object F`
+- 5. Create many new set elements to get the heap of `object F` back. These new elements are carefully constructed so that the `len` field representing the length of `NFT_SET_EXT_USERDATA` is exactly at the position of `object F->use`.
+- 6. Delete all the set elements we created in step 2 except `element X`. Now the `setelem->udata_len = 0xfb`(The original value is `0xa1-1=0xa0`)
+- 7. Dump all the set elements created in step 5. One of it will leak some usefull heap address. Now we will get heap address of `object E` and `object G`(objects created before and after `object F`)
+- 8. Delete all the set elements created in step 5 to free the heap of `object F` again.
+- 9. Spray memory filled with `addressof(object E)+0x80` to get the heap of `object F` back again.
+- 10. Dump set element `element X` created in step 2. We can leak `object E ->ops` because we overwrite `(*nft_set_ext_obj(ext))->key` by step 9.
+- 11. Delete all the `nft_object` we created in step 1 and spray ROP gadget to get all the heap back. Now we get two pointers by step 7 pointed to our ROP gadget.
+  
+  ```c
+  static int nf_tables_fill_setelem(struct sk_buff *skb,
+				  const struct nft_set *set,
+				  const struct nft_set_elem *elem)
+  ...
+  if (nft_set_ext_exists(ext, NFT_SET_EXT_OBJREF) &&
+	    nla_put_string(skb, NFTA_SET_ELEM_OBJREF,
+			   (*nft_set_ext_obj(ext))->key.name) < 0)
+		goto nla_put_failure;
+  ...
+  ```
+
+
+### Control RIP
+This part of the exploit is basically the same as CVE-2023-4569. The differences have been mentioned before. If you want to understand more details, read this [article](https://github.com/google/security-research/tree/master/pocs/linux/kernelctf/CVE-2023-4569_lts/docs/exploit.md).
+
+I control the RIP by the following steps:
+- 1. Create a chain for ROP(`chain X`)

```

Comment:
What size/type is this object and what cache is it allocated in?

Reply:
The size is 0x78 and it uses kmalloc-128

Reply:
Can you add this to the writeup?

Reply:
OK, I've added it in exploit.md.